zanata-war/src/main/java/org/zanata/rest/RestLimitingSynchronousDispatcher.java

```diff
@@ -45,18 +71,33 @@ public RestLimitingSynchronousDispatcher(
         this.processor = processor;
     }
 
+    HttpServletRequest getServletRequest() {
+        return ServletContexts.instance().getRequest();
+    }
+
     @Override
     public void invoke(final HttpRequest request, final HttpResponse response) {
 
+        //This is for REST request from browser where user is logged in.
         HAccount authenticatedUser = getAuthenticatedUser();
-        String apiKey = HeaderHelper.getApiKey(request);
+
+        String apiKey = HttpUtil.getApiKey(request);
 
         try {
-            // we are not validating api key but will rate limit any api key
-            if (authenticatedUser == null && Strings.isNullOrEmpty(apiKey)) {
-                response.sendError(
-                        Response.Status.UNAUTHORIZED.getStatusCode(),
-                        API_KEY_ABSENCE_WARNING);
+            if (authenticatedUser == null
+                    && Strings.isNullOrEmpty(apiKey)
+                    && !SecurityFunctions.canAccessRestPath(authenticatedUser,
+                            request.getHttpMethod(),
+                            request.getPreprocessedPath())) {
+
+                /**
+                 * Not using response.sendError because that will allow JBoss to
+                 * intercept and generate html page with the message. We want to
+                 * return a message string.
```

Comment:
To be clearer (assuming I've understood correctly), how about this?

	/**
	* Not using response.sendError because the app server will generate
	* an HTML page which includes the message. We want to return
	* the message string as is.
	*/